node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt

```diff
@@ -255,6 +273,55 @@ class DBTransactionStorage(private val database: CordaPersistence, cacheFactory:
         }
     }
 
+    /**
+     * returns a Paged Snapshot with give paging spec by querying transaction db
+     */
+    private fun pagedSnapshot(paging_: PageSpecification): TransactionStorage.Page {
+        val paging = if (paging_.pageSize == Integer.MAX_VALUE) {
+            paging_.copy(pageSize = Integer.MAX_VALUE - 1)
+        } else {
+            paging_
+        }
+        log.info("Paging spec for transaction snapshot: $paging")
+        // calculate total transactions where a page specification has been defined
+        var  totalTransactions = txStorage.content.size;
+
+        // pagination checks
+        if (!paging.isDefault) {
+            // pagination
+            if (paging.pageNumber < DEFAULT_PAGE_NUM) throw TransactionStorage.TransactionsQueryException("invalid page number ${paging.pageNumber} [page numbers start from $DEFAULT_PAGE_NUM]")
+            if (paging.pageSize < 1) throw TransactionStorage.TransactionsQueryException("invalid page size ${paging.pageSize} [minimum is 1]")
+            if (paging.pageSize > MAX_PAGE_SIZE) throw TransactionStorage.TransactionsQueryException("invalid page size ${paging.pageSize} [maximum is $MAX_PAGE_SIZE]")
+        }
+
+        //query
+        val session = currentDBSession()
+        val cb = session.criteriaBuilder
+        val criteriaQuery = cb.createQuery(DBTransaction::class.java)
+        val root = criteriaQuery.from(DBTransaction::class.java)
+        criteriaQuery.select(root)
+        criteriaQuery.orderBy(cb.desc(root.get<Instant>("timestamp")))
+        val query = session.createQuery(criteriaQuery)
+
+        query.firstResult = maxOf(0, (paging.pageNumber - 1) * paging.pageSize)
+        val pageSize = paging.pageSize + 1
+        query.maxResults = if (pageSize > 0) pageSize else Integer.MAX_VALUE // detection too many results, protected against overflow
+
+        // final pagination check (fail-fast on too many results when no pagination specified)
+        if (paging.isDefault && query.maxResults > DEFAULT_PAGE_SIZE) {
```

Comment:
this and the above assignment of Integer.MAX_VALUE can be consolidated to:
`query.maxResults = if (pageSize > 0) pageSize else DEFAULT_PAGE_SIZE`

this way the exception condition below can be eliminated (note that the exception message wasn't accurate because maxResults was just assigned a default value, Integer.MAX_VALUE, instead of being the actual result count)

Reply:
Thanks for catching that. They were some other issues as well in error handling. pushed changes to resolve